### PR TITLE
Add warnings related to Microsoft.NET.Sdk.WindowsDesktop

### LINKED ISSUE #866: Support Microsoft.NetCore.App.Wpf and Microsoft.NetCore.App.WinForms profiles in Microsoft.NET.Sdk.WidnowsDesktop
Support `Microsoft.NetCore.App.Wpf` and `Microsoft.NetCore.App.WinForms` profiles in `Microsoft.NET.Sdk.WidnowsDesktop` when `UseWpf` and/or `UseWindowsForms` is `true`

https://github.com/dotnet/sdk/pull/3259

### LINKED ISSUE #867: Show error when neither UseWpf nor UseWindowsForms is set to true
When neither `UseWpf` nor `UseWindowsForms` is set to `true`, `Microsoft.NET.Sdk.WindowsDesktop` should show a (suppressible) error

### PR DESCRIPTION
**Pull request by vatsan-madhavan**
Related: #3126 
This is also needed to completely fix https://github.com/dotnet/wpf/issues/866 - we still can't support support multitargeting `netcoreapp2.2`, and also to fix https://github.com/dotnet/wpf/issues/867

- Adds warning strings related to `Microsoft.NET.Sdk.WindowsDesktop`
- Adds a check to warn when `UseWpf` or `UseWindowsForms` is set without including `Microsoft.NET.Sdk.WindowsDesktop`

### CODE REVIEW

File: src/Tasks/Common/Resources/Strings.resx
Comment (peterhuene) [inline src/Tasks/Common/Resources/Strings.resx]: Nit:

```suggestion
    <value>NETSDK1107: Microsoft.NET.Sdk.WindowsDesktop is required to build Windows desktop applications. 'UseWpf' and 'UseWindowsForms' are not supported by the current SDK.</value>
```

File: src/Tasks/Common/Resources/Strings.resx
Comment (peterhuene) [inline src/Tasks/Common/Resources/Strings.resx]: Is this going to be used by a future change?

File: src/Tasks/Common/Resources/Strings.resx
Comment (peterhuene) [inline src/Tasks/Common/Resources/Strings.resx]: Is this going to be used by a future change?

**Review by peterhuene — APPROVED** [review]
Looks good, just a few questions and a nit.

File: src/Tasks/Common/Resources/Strings.resx
Comment (vatsan-madhavan) [reply src/Tasks/Common/Resources/Strings.resx]: Yes, as soon as this change flows into `dotnet/wpf`, we will consume this in `Microsoft.NET.Sdk.WindowsDesktop`

File: src/Tasks/Common/Resources/Strings.resx
Comment (vatsan-madhavan) [reply src/Tasks/Common/Resources/Strings.resx]: Yes, as soon as this change flows into `dotnet/wpf`, we will consume this in `Microsoft.NET.Sdk.WindowsDesktop`

**Comment by peterhuene** [comment]
Tests are failing with the new warning (treated as error).  Perhaps there are some test scenarios that need updating?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.targets]: It looks like `_MicrosoftNetSdkWindowsDesktop` isn't set by the WindowsDesktop SDK yet.  Until it is, this will block WindowsDesktop projects from building.  I think you need to set that property in the WindowsDesktop SDK first, flow that through to core-sdk and then to stage 0 in this repo, before committing this change.

Comment (vatsan-madhavan) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.targets]: I'm watching DARC flows (I triggered one earlier this morning from core-setup -> toolset exactly for this purpose: https://github.com/dotnet/toolset/pull/1177). Won't merge this until it's _very safe_ to do so.

Comment (vatsan-madhavan) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.targets]: Update global.json with latest core-sdk that contains support for `_MicrosoftNetSdkWindowsDesktop`

File: src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf
Comment (nguerrera) [inline src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf]: Something went wrong and things are resetting to English in some files. Can you try reverting all the xlf and let it regenerate? Also, if you have any insight on how this happened, let's add an issue to github.com/dotnet/xliff-tasks.

Comment (peterhuene) [reply src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf]: Nice catch as I missed this (unexpanded) file.  I wonder what caused this.

Comment (peterhuene) [reply src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf]: It happened to `src/Tasks/Common/Resources/xlf/Strings.tr.xlf` too.

Comment (vatsan-madhavan) [reply src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf]: Looking through the history, this happened in the very first commit. I'll redo the work and look through each file this time... and see if I can repro the problem along the way..

**Comment by vatsan-madhavan** [comment]
Anyone have ideas on why `Windows_NT_TestAsTools Build_Debug`  is failing ?

File: src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf
Comment (vatsan-madhavan) [reply src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf]: Fixed these.

**Comment by dsplaisted** [comment]
@wli3 Can you look at the TestAsTools failure here?  It looks like an issue with local tools and/or Json parsing.  (It's possible that the way the tool is being run is mixing up different pieces of the .NET Core SDK from stage 0 and from the repo in an incompatible way after stage 0 has been updated).

```
    System.TypeLoadException: Could not load type 'System.Text.Json.Serialization.JsonSerializer' from assembly 'System.Text.Json, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.
       at Microsoft.DotNet.ToolPackage.LocalToolsResolverCache.Save(IDictionary`2 restoredCommandMap)
       at Microsoft.DotNet.Tools.Tool.Install.LocalToolsResolverCacheExtensions.SaveToolPackage(ILocalToolsResolverCache localToolsResolverCache, IToolPackage toolDownloadedPackage, String targetFrameworkToInstall)
       at Microsoft.DotNet.Tools.Tool.Install.ToolInstallLocalCommand.Install(FilePath manifestFile)
       at Microsoft.DotNet.Tools.Tool.Install.ToolInstallLocalCommand.Execute()
       at Microsoft.DotNet.Tools.Tool.Install.ToolInstallCommand.Execute()
       at Microsoft.DotNet.Cli.DotNetTopLevelCommandBase.RunCommand(String[] args)
       at Microsoft.DotNet.Tools.Tool.ToolCommand.Run(String[] args)
       at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)
       at Microsoft.DotNet.Cli.Program.Main(String[] args)
    Errors
        F:\workspace\_work\1\s\src\Tests\Directory.Build.targets(56,5): error MSB3073: The command "dotnet tool install --local testSdkClean --version 3.0.100-ci --add-source F:\workspace\_work\1\s\artifacts\packages\Debug\NonShipping\" exited with code 1. [F:\workspace\_work\1\s\src\Tests\Microsoft.NET.Clean.Tests\Microsoft.NET.Clean.Tests.csproj]
```

**Comment by wli3** [comment]
@dsplaisted yes. there was a breaking change from s.t.json. and we get a mix of assembly versions.

**Comment by wli3** [comment]
update stage 0 should resolve it

File: global.json
Comment (wli3) [inline global.json]: @vatsan-madhavan I cannot push to this branch. Could you update this line to `3.0.100-preview7-012358`?

**Comment by wli3** [comment]
never mind, there should be a new breaking change. I need to push it though core-sdk to fix it

**Comment by wli3** [comment]
Here is the breaking change https://github.com/dotnet/corefx/pull/38120 cc @ahsonkhan

**Comment by wli3** [comment]
my PR is here https://github.com/dotnet/cli/pull/11532 however, after merging the change need to flow to core-sdk. and then we need to update the stage 0 in this PR to the new build core-sdk

**Comment by wli3** [comment]
@dsplaisted @livarcocc we need to wait for about a day to check this in. or we can consider ignore "Windows_NT_TestAsTools Build_Debug" if we need to get it in quick

**Comment by vatsan-madhavan** [comment]
@wli3 Windows_NT_TestAsTools Build_Debug is continuing to fail with the latest core-sdk in Stage 0. Any ideas when this will be fixed? I'm eager to get this change to get in since there is some preview 7 work in WPF that depends on these changes flowing through to core-sdk and then onto dotnet/wpf....

**Comment by wli3** [comment]
We have an CI issue that is blocking the dependency flow. I will add you to the thread

https://github.com/dotnet/toolset/pull/1207

**Comment by wli3** [comment]
@vatsan-madhavan meanwhile do you mind add me to your fork as contributor? I can push it through finish if i have the access to push to the branch

**Comment by nguerrera** [comment]
Normally you have push access to the PR source branch if PR has the checkbox to allow maintainers to do so (which it does by default). This allows this without maintaining collaborator lists in forks.

**Comment by vatsan-madhavan** [comment]
@wli3 just added you. if i unchecked the 'allow maintainers to edit..' checkbox, i have no memory of doing so. when i try to create another PR from my fork, the checkbox is definitely enabled by default as @nguerrera describes it (which means I must have unchecked it :( and forgotten all about it... ).

**Comment by wli3** [comment]
also added e6c78c75da7556b916da5d010aa0d976a46e6b19 to adjust runtime pack name change talked to @dsplaisted

**Comment by wli3** [comment]
@fadimounir https://dev.azure.com/dnceng/public/_build/results?buildId=224313&view=ms.vss-test-web.build-test-results-tab these new failing tests are in your area. Do you know them?

**Comment by fadimounir** [comment]
Looks like some update that is maybe breaking the logic around this line: https://github.com/dotnet/sdk/blob/fce5096afdec48b54dfd504f2a897933beb640d6/src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs#L68

Not all failing tests are R2R though...

**Comment by dsplaisted** [comment]
@wli3 From the output the It_publishes_the_project_with_a_refs_folder_and_correct_deps_file tests are failing because a new file has been added to the runtime, and the baseline needs to be updated:

> Unexpected compile libraries: System.Text.Encoding.CodePages.dll

The crossgen failures are failing with:

> /__w/1/s/artifacts/bin/Debug/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(245,5): error NETSDK1094: Unable to optimize assemblies for performance: a valid runtime package was not found. Either set the PublishReadyToRun property to false, or use a supported runtime identifier when publishing. [/__w/1/s/artifacts/tmp/Debug/It_creates_re---D21803AE/CrossgenTest3/CrossgenTest3.csproj]

That looks like it's probably also related to the runtime pack rename

**Comment by dsplaisted** [comment]
@vatsan-madhavan This should hopefully be unblocked if you rebase it on top of #3333, which was just merged

**Comment by vatsan-madhavan** [comment]
The build finally passed! 

Appreciate it if someone can merge at earliest convenience!

**Pull request merged**
